### ISSUE TITLE
Intermittently failing to send events or properties

### ISSUE THREAD

**Issue by vibinr** [mixpanel/mixpanel-android#372]
Mixpanel is randomly missing out on some events, super properties etc. This is what my current flow looks like:

I do `MixpanelAPI.getInstance()` in my Application class and use this in all the needed activities. `MixpanelAPI.flush()` is being called in onDestroy of all these activities. 

Am I missing something?

One more question, can I call `mixpanelAPI.getPeople().setMap()` and `mixpanelAPI.registerSuperPropertiesMap()` without actually firing an event? I do this at few places to update existing super/people properties.

Thanks for the awesome lib!

**Comment by yinfeiru** [comment]
@vibinr 

Calling `MixpanelAPI.flush()` in `onDestroy` doesn't always work since `onDestroy` only gets called when the app is shut down. It'd be better if you flush the queue whenever the user backgrounds the app.

You don't have to fire any event to make `mixpanelAPI.getPeople().setMap()`  and `m`ixpanelAPI.registerSuperPropertiesMap()` work.

**Comment by patedit** [comment]
Hi @vibinr Is this still a problem for you?

**Comment by vibinr** [comment]
@patedit it's a lot better now and happens only once in 100 users or so.

**Comment by patedit** [comment]
That's still unacceptable for us. How are you confirming the lose of events? Do you have some steps to reproduce the issue?

**Comment by vibinr** [comment]
We weren't able to reproduce this ourselves. We noticed this issue in our previous onboarding flow, where user had to go through two steps - uploadStore() and editStore(). These routines looked like this:

```
private void uploadStore() {
    // upload store to server, with user filled params, and empty phone number, empty email address
	Map<String, Object> superProps = getMixpanelShopSuperProps();
	Map<String, Object> pplProps = getMixpanelShopPplProps();

	mixpanelAPI.getPeople().setMap(pplProps);
	mixpanelAPI.registerSuperPropertiesMap(superProps);
	// this gets fired correctly, with empty phone number, empty email address
	mixpanelAPI.track(AnalyticsConstants.Events.CREATE_SHOP);
}

private void editStore() {
	// ask user to fill in mandatory fields like phone number, email address
	// and send updated store info to server
	Map<String, Object> superProps = getMixpanelShopSuperProps();
	Map<String, Object> pplProps = getMixpanelShopPplProps();

	mixpanelAPI.getPeople().setMap(pplProps);
	mixpanelAPI.registerSuperPropertiesMap(superProps);
       // no .track() called here
}
```

After user goes through this flow, we weren't able to see updated properties (after editStore() is called) on MIxpanel. That is, it showed phone number and email addresses as empty for quite a few users. Though our database is updated fine and we can see users' registered phone number/email address.

We have made a lot of changes to our onboarding lately and haven't come across this issue again 😄

**Comment by patedit** [comment]
Thanks @vibinr for the clarification! It'd be useful to see `getMixpanelShopSuperProps` and `getMixpanelShopPplProps` implementations to see how you are gathering those maps. Also, a critical part is when and how the number/email are persisted. 
Since it seems that after changing the onboarding experience the problem is solved I am going to close this issue but, please, feel free to open a new ticket is the problem arises again. Thanks!
